src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs

```diff
@@ -66,21 +63,25 @@ internal sealed class ProjectCodeModelFactory : IProjectCodeModelFactory
             _visualStudioWorkspace.WorkspaceChanged += OnWorkspaceChanged;
         }
 
-        private System.Threading.Tasks.Task ProcessNextDocumentBatchAsync(
+        private Task ProcessNextDocumentBatchAsync(
             ImmutableArray<DocumentId> documentIds, CancellationToken cancellationToken)
         {
             foreach (var documentId in documentIds)
             {
                 // Now, enqueue foreground work to actually process these documents in a serialized and incremental
                 // fashion.  FireEventsForDocument will actually limit how much time it spends firing events so that it
                 // doesn't saturate  the UI thread.
-                _notificationService.RegisterNotification(
-                    () => FireEventsForDocument(documentId),
-                    _listener.BeginAsyncOperation("CodeModelEvent"),
-                    cancellationToken);
+                _threadingContext.JoinableTaskFactory.RunAsync(async () =>
+                {
+                    using var _ = _listener.BeginAsyncOperation("CodeModelEvent");
+
+                    await Task.Delay(15, cancellationToken).ConfigureAwait(false);
+                    await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                    FireEventsForDocument(documentId);
+                });
```

Comment:
💭 This loop posts more messages to the UI thread than it used to. It might be better to place the `foreach` inside the call to `RunAsync` and then use `SharedStopwatch` to add await `Task.Yield` after a sequence of events longer than some allowable slice.

Reply:
wfm